### DISCUSSION TITLE
API URL generation faster

### DISCUSSION THREAD

**Discussion by semaf** [YOURLS/YOURLS#3778] [Q&A]
Hey,

I am shortning with curl, sample code of yourls, the urls. It takes 40-50 seconds for 30 urls. My list has also short urls which are already shorten. 

Anyone who is doing it much faster without DB requests etc?

**Comment by logusivam** [comment]
40–50 seconds for 30 URLs is far too slow — that's roughly 1.5 seconds per URL, which points to a specific set of fixable bottlenecks, not a fundamental YOURLS limitation. With the right approach you can get 30 URLs down to under 2 seconds total.

---

## Why it is slow — the three culprits

### 1. Sequential curl requests (biggest factor)
The standard YOURLS sample code sends one curl request, waits for the full response, then sends the next. With network latency of even 50ms per request, 30 sequential requests adds up fast — and if your server is doing remote title fetching for each URL, that adds another 1–2 seconds per URL while PHP waits for Amazon/Google/etc. to respond.

### 2. Title fetching for each URL
By default YOURLS makes an outbound HTTP request to fetch the page title of every URL being shortened. For 30 URLs that is 30 additional outbound HTTP calls happening server-side, each waiting for a remote server to respond.

### 3. Already-shortened URLs triggering a full lookup cycle
When a URL already exists in the database, YOURLS still runs the full request pipeline (auth, sanitisation, title fetch, duplicate check) before returning the existing short URL. There is no early-exit fast path in the default API.

---

## Fix 1 — Use curl_multi to send all requests in parallel (biggest win)

Instead of sending 30 requests one after another, send them all simultaneously and collect results as they complete. This alone typically reduces 30 sequential requests to the time of the single slowest one.

```php
<?php

$yourls_site   = 'https://yourdomain.com';
$yourls_token  = 'your_signature_token';

$urls = [
    'https://example.com/page1',
    'https://example.com/page2',
    // ... up to 30 URLs
];

function yourls_shorten_parallel( $urls, $yourls_site, $token ) {
    $mh      = curl_multi_init();
    $handles = [];
    $results = [];

    foreach ( $urls as $i => $url ) {
        $ch = curl_init();
        curl_setopt_array( $ch, [
            CURLOPT_URL            => $yourls_site . '/yourls-api.php',
            CURLOPT_POST           => true,
            CURLOPT_POSTFIELDS     => http_build_query([
                'signature' => $token,
                'action'    => 'shorturl',
                'url'       => $url,
                'format'    => 'json',
                'title'     => ' ',    // pass a space to skip server-side title fetch
            ]),
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT        => 10,
            CURLOPT_CONNECTTIMEOUT => 5,
        ]);
        $handles[ $i ] = $ch;
        curl_multi_add_handle( $mh, $ch );
    }

    // Execute all requests simultaneously
    $running = null;
    do {
        curl_multi_exec( $mh, $running );
        curl_multi_select( $mh );
    } while ( $running > 0 );

    // Collect results
    foreach ( $handles as $i => $ch ) {
        $response      = curl_multi_getcontent( $ch );
        $data          = json_decode( $response, true );
        $results[ $i ] = [
            'original'  => $urls[ $i ],
            'shorturl'  => $data['shorturl']  ?? null,
            'status'    => $data['status']    ?? 'error',
            'message'   => $data['message']   ?? 'unknown error',
        ];
        curl_multi_remove_handle( $mh, $ch );
        curl_close( $ch );
    }

    curl_multi_close( $mh );
    return $results;
}

$start   = microtime( true );
$results = yourls_shorten_parallel( $urls, $yourls_site, $yourls_token );
$elapsed = microtime( true ) - $start;

foreach ( $results as $r ) {
    echo $r['original'] . ' → ' . ( $r['shorturl'] ?? $r['message'] ) . PHP_EOL;
}

echo PHP_EOL . 'Done in ' . round( $elapsed, 2 ) . 's' . PHP_EOL;
```

> ⚠️ If you are shortening a very large list (hundreds of URLs), batch into groups of 20–30 parallel requests rather than firing all at once — this avoids overwhelming your server and hitting PHP/Apache connection limits.

---

## Fix 2 — Suppress server-side title fetching

Notice `'title' => ' '` in the curl options above. Passing any non-empty title value tells YOURLS to use that as the title instead of making an outbound HTTP request to fetch it. Passing a space is the minimal way to skip the fetch entirely. This alone can save 1–2 seconds per URL when the remote server is slow to respond.

If you want no title stored at all, pass an empty string — but note that some YOURLS versions treat an empty title as "please fetch it", so a single space is safer.

---

## Fix 3 — Pre-filter already-shortened URLs client-side

If your list contains URLs that are already shortened, checking them against YOURLS before sending wastes an API round-trip. Keep a local cache of URLs you've already processed:

```php
<?php

// Simple file-based cache — use Redis or a DB table for large lists
$cache_file = __DIR__ . '/shortened_cache.json';
$cache      = file_exists( $cache_file )
              ? json_decode( file_get_contents( $cache_file ), true )
              : [];

$urls_to_shorten = [];
$already_done    = [];

foreach ( $urls as $url ) {
    if ( isset( $cache[ $url ] ) ) {
        $already_done[ $url ] = $cache[ $url ];
    } else {
        $urls_to_shorten[] = $url;
    }
}

// Only send the ones not already cached
if ( $urls_to_shorten ) {
    $new_results = yourls_shorten_parallel( $urls_to_shorten, $yourls_site, $yourls_token );
    foreach ( $new_results as $r ) {
        if ( $r['shorturl'] ) {
            $cache[ $r['original'] ] = $r['shorturl'];
        }
    }
    file_put_contents( $cache_file, json_encode( $cache, JSON_PRETTY_PRINT ) );
}

$all_results = array_merge( $already_done, array_column( $new_results ?? [], 'shorturl', 'original' ) );
```

For large lists this is the single biggest time saver — URLs already in YOURLS return an existing short URL anyway, so you are paying an API round-trip for zero benefit.

---

## Fix 4 — Use the signature token, not username/password auth

If you are using `username` + `password` in your API calls instead of the `signature` token, switch to signature. Password-based auth adds a bcrypt verification step on every single request. The signature token skips this entirely.

Get your token from **Admin → Tools → Secure passwordless API call**.

```php
// Slow — bcrypt on every request
'username' => 'admin',
'password' => 'yourpassword',

// Fast — token lookup only
'signature' => 'your_signature_token',
```

---

## Fix 5 — Add a database index if `yourls_log` is large

If you have millions of rows in `yourls_log` (common on busy installs), even the duplicate-URL check query can be slow. Ensure this index exists:

```sql
ALTER TABLE yourls_url ADD INDEX IF NOT EXISTS idx_url (url(255));
```

This makes the "does this URL already exist?" lookup a fast index scan instead of a full table scan.

---

## Expected performance after these fixes

| Approach | Time for 30 URLs |
|----------|-----------------|
| Sequential curl (original) | 40–50 seconds |
| Parallel curl only | 2–5 seconds |
| Parallel curl + skip title fetch | 1–3 seconds |
| Parallel curl + client-side cache + skip title fetch | < 1 second for cached URLs |

---

## Summary — apply in this order

1. **Switch to `curl_multi` parallel requests** — this alone will cut your time by 90%.
2. **Pass a title to skip server-side title fetching** — saves 1–2s per URL.
3. **Use signature token auth** — eliminates bcrypt overhead per request.
4. **Cache already-shortened URLs client-side** — avoid redundant API calls entirely.
5. **Add a DB index on `yourls_url.url`** — speeds up duplicate detection queries.

Hope this helps — if it answers your question, would you mind clicking **"Mark as answer"** so others searching for the same thing can find it quickly?
